docs/anti-patterns.mdx

```diff
@@ -4,35 +4,152 @@ title: Cadence Anti-Patterns
+// should not be accessible in this function
+// BAD
+
+pub fun transferNFT(id: UInt64, owner: AuthAccount) {
+    assert(owner(id) == owner.address)
+
+    transfer(id)
+
+    // Sneakily borrow a reference to the user's Flow Token Vault
+    // and withdraw a bit of FLOW
+    // BAD
+    let vaultRef = owner.borrow<&FlowToken.Vault>(/storage/flowTokenVault)!
+    let stolenTokens <- vaultRef.withdraw(amount: 0.1)
+
+    // deposit the stolen funds in the contract owners vault
+    // BAD
+    contractVault.deposit(<-stolenTokens)
+}
+...
+```
+
+### Solution
+
+Projects should find other ways to authenticate users, such as using resources and capabilities as authentication objects.
+They should also expect to perform most storage and linking operations within transaction bodies
+instead of in a contract utility functions.
+
+There are some scenarios where using an `AuthAccount` object is necessary, such as a cold storage multi-sig,
+but those cases are extremely rare and `AuthAccount` usage should still be avoided unless absolutely necessary.
 
-If that resource has functions that emit events, any account can create an instance of that resource and emit those events.
+## Auth References should be avoided
 
-If those events are meant to indicate actions taken using a single instance of that resource (for example an admin object, switchboard, or registry), or instances created through a particular workflow, then this will make event log search and management harder because events from other instances may be mixed in with the ones that you wish to examine.
+### Problem
 
-To fix this, if there should be only a single instance of the resource it should be created and `link()`ed to a public path in an admin account's storage during the contracts's initializer.
+[`auth` references](/language/references) should be avoided unless necessary:
+An Auth Reference allows downcasting any restricted type to its unrestricted type.
+The parent type could expose functionality that was not intended to be exposed.
+If the auth keyword is used on local variables they will be references.
+References are ephemeral and cannot be stored.
+This prevents any reference casting to be stored under account storage.
+Additionally, if the `auth` keyword is used to store a public capability, serious harm
+could be happen since the value could be downcasted to a type
+that has functionality and values altered.
 
-### Named Value Fields Are Preferable
+### Example
 
-Where contracts, resources, and scripts all have to repeatedly refer to the same constant values, entering these manually is a potential source of error. These are known as [magic values](<https://en.wikipedia.org/wiki/Magic_number_(programming)>).
+A commonly seen pattern in NFT smart contracts is including a public borrow function
+that borrows an auth reference to an NFT (eg. [NBA Top Shot](https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L889-L906)).
```

Comment:
Link targets the master branch, which might change. Instead, link to the current commit, i.e `
```suggestion
that borrows an auth reference to an NFT (eg. [NBA Top Shot](https://github.com/dapperlabs/nba-smart-contracts/blob/95fe72b7e94f43c9eff28412ce3642b69dcd8cd5/contracts/TopShot.cdc#L889-L906)).
```

(Tip: Pressing `y` in GitHub changes the URL from branch to commit)